subsys/usb/host/usbh_shell.c

```diff
@@ -0,0 +1,627 @@
+
+#define FOOBAZ_VREQ_OUT		0x5b
+#define FOOBAZ_VREQ_IN		0x5c
+
+USBH_CONTROLLER_DEFINE(uhs_ctx, DEVICE_DT_GET(DT_NODELABEL(zephyr_uhc0)));
+
+const static struct shell *ctx_shell;
+
+static void print_dev_desc(const struct shell *sh,
+			   const struct usb_device_descriptor *const desc)
+{
+	shell_print(sh, "bLength\t\t\t%u", desc->bLength);
+	shell_print(sh, "bDescriptorType\t\t%u", desc->bDescriptorType);
+	shell_print(sh, "bcdUSB\t\t\t%x", desc->bcdUSB);
+	shell_print(sh, "bDeviceClass\t\t%u", desc->bDeviceClass);
+	shell_print(sh, "bDeviceSubClass\t\t%u", desc->bDeviceSubClass);
+	shell_print(sh, "bDeviceProtocol\t\t%u", desc->bDeviceProtocol);
+	shell_print(sh, "bMaxPacketSize0\t\t%u", desc->bMaxPacketSize0);
+	shell_print(sh, "idVendor\t\t%x", desc->idVendor);
+	shell_print(sh, "idProduct\t\t%x", desc->idProduct);
+	shell_print(sh, "bcdDevice\t\t%x", desc->bcdDevice);
+	shell_print(sh, "iManufacturer\t\t%u", desc->iManufacturer);
+	shell_print(sh, "iProduct\t\t%u", desc->iProduct);
+	shell_print(sh, "iSerial\t\t\t%u", desc->iSerialNumber);
+	shell_print(sh, "bNumConfigurations\t%u", desc->bNumConfigurations);
+}
+
+static void print_cfg_desc(const struct shell *sh,
+			   const struct usb_cfg_descriptor *const desc)
+{
+	shell_print(sh, "bLength\t\t\t%u", desc->bLength);
+	shell_print(sh, "bDescriptorType\t\t%u", desc->bDescriptorType);
+	shell_print(sh, "wTotalLength\t\t%x", desc->wTotalLength);
+	shell_print(sh, "bNumInterfaces\t\t%u", desc->bNumInterfaces);
+	shell_print(sh, "bConfigurationValue\t%u", desc->bConfigurationValue);
+	shell_print(sh, "iConfiguration\t\t%u", desc->iConfiguration);
+	shell_print(sh, "bmAttributes\t\t%02x", desc->bmAttributes);
+	shell_print(sh, "bMaxPower\t\t%u mA", desc->bMaxPower * 2);
+}
+
+static void print_desc(const struct shell *sh, const struct net_buf *const buf)
+{
+	struct usb_desc_header *head = (void *)buf->data;
+
+	switch (head->bDescriptorType) {
+	case USB_DESC_DEVICE: {
+		struct usb_device_descriptor *desc = (void *)buf->data;
+
+		desc->bcdUSB = sys_le16_to_cpu(desc->bcdUSB);
+		desc->idVendor = sys_le16_to_cpu(desc->idVendor);
+		desc->idProduct = sys_le16_to_cpu(desc->idProduct);
+		desc->bcdDevice = sys_le16_to_cpu(desc->bcdDevice);
+		print_dev_desc(sh, desc);
+		break;
+	}
+	case USB_DESC_CONFIGURATION: {
+		struct usb_cfg_descriptor *desc = (void *)buf->data;
+
+		desc->wTotalLength = sys_le16_to_cpu(desc->wTotalLength);
+		print_cfg_desc(sh, desc);
+		break;
+	}
+	default:
+		shell_hexdump(ctx_shell, buf->data, buf->len);
+		break;
+	}
+}
+
+static int bazfoo_request(struct usbh_contex *const ctx,
+			  struct uhc_transfer *const xfer,
+			  int err)
+{
+	const struct device *dev = ctx->dev;
+
+	shell_info(ctx_shell, "host: transfer finished %p, err %d", xfer, err);
+
+	while (!k_fifo_is_empty(&xfer->done)) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&xfer->done, K_NO_WAIT);
+		if (buf) {
+			if (xfer->ep == USB_CONTROL_EP_IN) {
+				print_desc(ctx_shell, buf);
```

Comment:
This code assumes that all control requests are reading descriptors which is not true for `FOOBAZ_VREQ_IN`. To solve this, the request context should be somehow preserved. Just a FIXME comment for now seems sufficient though.

Reply:
Added comment